ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -13,6 +13,151 @@ namespace ext
 namespace MitsubaLoader
 {
 
+static uint64_t rgb32f_to_rgb19e7(const uint32_t _rgb[3])
+{
+	constexpr uint32_t mantissa_bitlen = 19u;
+	constexpr uint32_t exp_bitlen = 7u;
+	constexpr uint32_t exp_bias = 63u;
+
+	constexpr uint32_t mantissa_bitlen_f32 = 23u;
+	constexpr uint32_t exp_bitlen_f32 = 8u;
+	constexpr uint32_t exp_bias_f32 = 127u;
+
+	constexpr uint32_t mantissa_len_diff = mantissa_bitlen_f32 - mantissa_bitlen;
+	constexpr uint32_t exp_bias_diff = exp_bias_f32 - exp_bias;
+
+	uint64_t rgb19e7 = 0ull;
+	for (uint32_t i = 0u; i < 3u; ++i)
+	{
+		uint64_t mantissa = _rgb[i] & 0x7fffffu;
+		mantissa >>= mantissa_len_diff;
+		mantissa <<= (mantissa_bitlen * i);
+		rgb19e7 |= mantissa;
+	}
+	uint64_t exp = (_rgb[0] >> mantissa_bitlen_f32) & 0xffu;
+	exp -= exp_bias_diff;
+	rgb19e7 |= (exp << (3u * mantissa_bitlen));
+
+	return rgb19e7;
+}
+
+static bsdf::SBSDFUnion bsdfNode2bsdfStruct(const CElementBSDF* _node)
+{
+	bsdf::SBSDFUnion retval;
+	uint32_t bitfields = _node->type;
+	switch (_node->type)
+	{
+	case CElementBSDF::Type::DIFFUSE:
+		_IRR_FALLTHROUGH;
```

Comment:
this is dangerous, nobody knows what's going to be in `_node->diffuse.alpha.value.type`

Reply:
this applies to almost every single fallthrough in this switch

Reply:
ah, good point